programs/dxfread.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.48.5.
.TH DXFREAD "1" "November 2023" "dxfread 0.13.0" "User Commands"
.SH NAME
dxfread \- manual page for dxfread 0.13.0
.SH SYNOPSIS
.B dxfread
[\fI\,OPTION\/\fR]... \fI\,DXFFILE\/\fR
.SH DESCRIPTION
Reads the DXF into some optional output format to stdout or some file,
and prints error, success or verbose internal progress to stderr.
.TP
\fB\-v[0\-9]\fR, \fB\-\-verbose\fR [0\-9]
verbosity
.TP
\fB\-O\fR fmt,  \fB\-\-format\fR fmt
fmt: DXF, DXFB, JSON, GeoJSON
.TP
Planned output formats:
YAML, XML/OGR, GPX, SVG, PS
.TP
\fB\-o\fR outfile
also defines the output fmt. Default: stdout
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.PP
GNU LibreDWG online manual: <https://www.gnu.org/software/libredwg/>
.SH "SEE ALSO"
The full documentation for
.B dxfread
is maintained as a Texinfo manual.  If the
.B info
and
.B dxfread
programs are properly installed at your site, the command
.IP
.B info LibreDWG
.PP
should give you access to the complete manual.


NEWS
GNU LibreDWG NEWS -- history of user-visible changes.  -*-indented-text-*-
Copyright (C) 2024 Free Software Foundation, Inc.
See the end for copying conditions.

LibreDWG version 0.13.3 - 2023-02-26 - beta:

Minor features:
  * Promoted SUN object to stable
Minor bugfixes:
  * Fix make dist, checking missing dwg2ps.1 and dwgadd.1. Check for missing
    libps. (#941)
  * Improved BLOCKSTRETCHACTION a bit
  * Improved dxf import for OLE2FRAME.data
  * Fix some fuzzing errors: ossfuzz 66843, 66835, 66639
  * Various unknown fixes, esp. for out-of-tree builds
  * Improved wrong DWG_MAX_OBJSIZE, which prevented
    very large 3DSOLID's to encode, like STEERING.dwg from_autocad_r13
  * Improved support for beta versions, no logic holes (#908)

LibreDWG version 0.13.2 - 2023-02-10 - beta:

The source releases 0.13 and 0.13.1 missed test/xmlsuite and bindings/python,
due to a broken automated release process.

Minor bugfixes:
  * Fix make dist, checking missing test/xmlsuite and bindings/python.
  * Fix the release CI process, install the missing python deps.
  * add build-aux/install-libxslt helper

LibreDWG version 0.13.1 - 2023-02-10 - beta:

Minor features:
  * dwgread now supports --fmt=minJSON or --file=XXX.min.json
    to emit minimized json without whitespace. Use eg.
      jq . XXX.min.json
    to unminimize it. Saves about 50% size.
Minor bugfixes:
  * Fixed DXF XRECORD xdata values and sizes (wrong binary encoding in outdxf,
    wrong xdata_size in indxf)
  * Fixed indxf for stable ASSOCGEOMDEPENDENCY object.
  * Fixed decoding handles with sizes > 4, now capped at 32bit.
    There were apparently undocumented handle bits in the 5th byte.
  * Fixed decoding preR13 entities with an offset of 1, one unknown byte.
    Fixes one r2.6 TEXT example.
  * Improve PROXY fields and DXF codes
  * Changed MLINESTYLE.line.lt fields (#924)
  * gcc-14 warning fixes (#931)
  * Fixed LWPOLYLINE.PLINEGEN flag
  * Fixed SummaryInfo.props[].tag and value types to T16
  * Change DIMSTYLE.DIMTXTDIRECTION dxf code from 295 to 294. (#835)
    Undocumented, ODA has it as 295, netDXF as 294. on indxf accept both.
  * Fix encode ERROR: Wrong object size on MS adjust on large objects.
  * Fix encode downconvert_DIMSTYLE, when the eed already had one
    member, AcadAnnotative.
  * Fixed preR13 generated internal handles, esp. for the control table records,
    by updating the internal HANDSEED. Which fixes the ERROR: Duplicate handle X
    already points to object X.

LibreDWG version 0.13 - 2023-02-04 - beta:

Major features:
  * Added pre-R13 DWG decode support for all old DWG formats
    with Michal Josef Špaček. DXF and many down-conversions not yet.
  * Add support for VS2019 (msvc with ucrt) and clang-cl cross-compilation
    to Windows. (it is not error compatible to msvc, msvc errors much more).
    Also cross-compile to mips32be and ppc64be.
    Tested ok with tcc and a patched chibicc. pcc and compcert not yet.
  * Added read support on bigendian platforms, write not yet.
  * Full support for en-/decoding foreign codepages, also the MIF
    \M=1-5xxxx format. (With and without iconv)
  * Added WMF and PNG support to dwgbmp (#824)
  * Identified the R13-2000 ODA Unknown section as ObjFreeSpace (#860).
  * Fix many minor ODA roundtrip errors, leading to less recovery mode
    dialogs.
  * Encode now support various r13-r2000 section orderings, as in the original.
  * Fixed dwgadd (Dmitry Sinyavin #388, #415)
  * Add --with-dxf-precision=(rfc,max,16,6) and
    --with-geojson-precision=(rfc,max,16,6) build options. (#810)
    The GeoJSON default is now back from the RFC7946 recommended 6 to 16 (=max),
    leading to much larger and much more precise files.
  * Split arcs/circles into line segments with GeoJSON (#520).
  * libredwg is on vcpkg
  * Added a fixed [ODA Spec diff](doc/ODA-5.4.2-libredwg-diff.pdf)
  * New authors: Michal Josef Špaček (preR13 DWG formats),
    Dmitry Sinyavin (dwgadd fixes).
Minor features:
  * Added a third *int alloced arg to dwg_dynapi_handle_name()
  * Renamed HEADER.unknown_s[] fields. (Michal Spacek #412)
  * Fixed RENDERENVIRONMENT.fog_color, seperated into extra rgb fields
  * Now stable: RENDERENVIRONMENT
  * json contains and can import now also SecondHeader and R2007_Header sections.
  * Better logging.
Minor bugfixes:
  * Fixed MESH roundtrips, added 2 unknown bits.
  * Fixed wrong r2007 thumbnail overflow errors (#824)
  * Fixed HEADER.PLINEGEN pre-r13 (Michal Spacek #408)
  * Fixed HEADER.OSMODE pre-r13 (Michal Spacek #409)
  * Fixed LAYER.plotflag (#509)
  * Remove APPID.EXEMPT_FROM_CAD_STANDARDS when down-encoding to r2000. (#817)
  * Create an VX_CONTROL and an empty VX table record when down-encoding to
    r2000. (#811)
  * Fixed all T16 texts in the SummaryInfo section <r2007.
  * Fixed many more minor fuzzer errors with illegal input, detected by the
    oss-fuzz service and others. More bounds and type checks, null-derefs,
    double-free, missing zero-termination in fuzzer inputs.

    Fixed GH fuzzing issues 911, 901, 899, 890, 880, 881, 859, 850,
    848, 708, 707, 704, 697, 696, 695, 694, 692, 677, 616, 615, 594,
    581, 578, 528, 527, 524, 518, 497, 494, 493, 492, 491, 490, 489,
    488, 487, 486, 485, 484, 425, 361, 360, 356.

    Fixed oss-fuzz issues 66111, 66046, 65937, 65481, 65366, 65359,
    65357 65355, 65353, 65346, 65125, 64751, 64506, 64318, 64118,
    64052, 63919, 63895 63824, 63814, 63803, 63799, 63798, 63791,
    63776, 63761, 63760, 63690, 63629 63537, 63483, 63463, 63137,
    63076, 63057, 62973, 62718, 62561, 62461, 62367 62363, 62132,
    61964, 61901, 61818, 61795, 61741, 61735, 61593, 61497, 61494
    60993, 60980, 60902, 60600, 60598, 60216, 60067, 60008, 59899,
    59847, 59826 59763, 59732, 59633, 59589, 59571, 59561, 59544,
    59528, 59523, 59515, 59428 59388, 59124, 59118, 59097, 59089,
    58974, 58927, 58915, 58904, 58649, 58564 58498, 58473, 58443,
    58334, 58306, 58143, 58135, 58082, 58019, 58008, 57981 57969,

programs/dxfread.c
/*****************************************************************************/
/*  LibreDWG - free implementation of the DWG file format                    */
/*                                                                           */
/*  Copyright (C) 2009-2023 Free Software Foundation, Inc.                   */
/*  Copyright (C) 2010 Thien-Thi Nguyen                                      */
/*                                                                           */
/*  This library is free software, licensed under the terms of the GNU       */
/*  General Public License as published by the Free Software Foundation,     */
/*  either version 3 of the License, or (at your option) any later version.  */
/*  You should have received a copy of the GNU General Public License        */
/*  along with this program.  If not, see <http://www.gnu.org/licenses/>.    */
/*****************************************************************************/

/*
 * dxfread.c: read a DXF file, print verbose logging, and output to
 *            various formats.
 * written by Felipe Castro
 * modified by Felipe Corrêa da Silva Sances
 * modified by Rodrigo Rodrigues da Silva
 * modified by Thien-Thi Nguyen
 * modified by Reini Urban
 */

#include "../src/config.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
// strings.h or string.h
#ifdef AX_STRCASECMP_HEADER
#  include AX_STRCASECMP_HEADER
#endif
#include "my_getopt.h"
#ifdef HAVE_VALGRIND_VALGRIND_H
#  include <valgrind/valgrind.h>
#endif

#include "dwg.h"
#include "common.h"
#include "bits.h"
#ifndef DISABLE_DXF
#  include "out_dxf.h"
#  ifndef DISABLE_JSON
#    include "out_json.h"
#  endif
#endif

#if !defined HAVE_STRCASECMP
int strcasecmp (const char *a, const char *b);
#endif

static int opts = 1;

static int
usage (void)
{
#ifndef DISABLE_DXF
  printf ("\nUsage: dxfread [-v[0-9]] [-O FMT] [-o OUTFILE] [DXFFILE|-]\n");
#else
  printf ("\nUsage: dxfread [-v[0-9]] [DXFFILE|-]\n");
#endif
  return 1;
}
static int
opt_version (void)
{
  printf ("dxfread %s\n", PACKAGE_VERSION);
  return 0;
}
static int
help (void)
{
  printf ("\nUsage: dxfread [OPTION]... DXFFILE\n");
  printf ("Reads the DXF into some optional output format to stdout or some "
          "file,\n"
          "and prints error, success or verbose internal progress to stderr.\n"
          "\n");
#ifdef HAVE_GETOPT_LONG
  printf ("  -v[0-9], --verbose [0-9]  verbosity\n");
#  ifndef DISABLE_DXF
#    ifndef DISABLE_JSON
  printf ("  -O fmt,  --format fmt     fmt: DXF, DXFB, JSON, GeoJSON\n");
#    else
  printf ("  -O fmt,  --format fmt     fmt: DXF, DXFB\n");
#    endif
  printf ("           Planned output formats:  YAML, XML/OGR, GPX, SVG, PS\n");
  printf ("  -o outfile                also defines the output fmt. Default: "
          "stdout\n");
#  endif
  printf ("           --help           display this help and exit\n");
  printf ("           --version        output version information and exit\n"
          "\n");
#else
  printf ("  -v[0-9]     verbosity\n");
#  ifndef DISABLE_DXF
#    ifndef DISABLE_JSON
  printf ("  -O fmt      fmt: DXF, DXFB, JSON, GeoJSON\n");
#    else
  printf ("  -O fmt      fmt: DXF, DXFB\n");
#    endif
  printf (
      "              Planned output formats:  YAML, XML/OGR, GPX, SVG, PS\n");
  printf ("  -o outfile  also defines the output fmt. Default: stdout\n");
#  endif
  printf ("  -h          display this help and exit\n");
  printf ("  -i          output version information and exit\n"
          "\n");
#endif
  printf ("GNU LibreDWG online manual: "
          "<https://www.gnu.org/software/libredwg/>\n");
  return 0;
}

int
main (int argc, char *argv[])
{
  int i = 1;
  int error;
  Dwg_Data dwg;
  const char *fmt = NULL;
  const char *outfile = NULL;
  int has_v = 0;
  int force_free = 0;
  int c;
#ifdef HAVE_GETOPT_LONG
  int option_index = 0;
  static struct option long_options[]
      = { { "verbose", 1, &opts, 1 }, // optional
          { "format", 1, NULL, 'O' },   { "file", 1, NULL, 'o' },
          { "help", 0, NULL, 0 },       { "version", 0, NULL, 0 },
          { "force-free", 0, NULL, 0 }, { NULL, 0, NULL, 0 } };
#endif

  if (argc < 2)
    return usage ();

  while (1)
    {
#ifdef HAVE_GETOPT_LONG
      c = getopt_long (argc, argv, "v::O:o:h", long_options, &option_index);
#else
      c = getopt (argc, argv, "v::O:o:hi");
#endif
      if (c == -1)
        break;
      switch (c)
        {
        case ':': // missing arg
          if (optarg && !strcmp (optarg, "v"))
            {
              opts = 1;
              has_v = 1;
              break;
            }
          fprintf (stderr, "%s: option '-%c' requires an argument\n", argv[0],
                   optopt);
          break;
#ifdef HAVE_GETOPT_LONG
        case 0:
          /* This option sets a flag */
          if (!strcmp (long_options[option_index].name, "verbose"))
            {
              if (opts < 0 || opts > 9)
                return usage ();
#  if defined(USE_TRACING) && defined(HAVE_SETENV)
              {
                char v[2];
                *v = opts + '0';
                *(v + 1) = 0;
                setenv ("LIBREDWG_TRACE", v, 1);
              }
#  endif
              has_v = 1;
              break;
            }
          if (!strcmp (long_options[option_index].name, "version"))
            return opt_version ();
          if (!strcmp (long_options[option_index].name, "help"))
            return help ();
          if (!strcmp (long_options[option_index].name, "force-free"))
            force_free = 1;
          break;
#else
        case 'i':
          return opt_version ();
#endif
        case 'O':
          fmt = strdup (optarg);
          break;
        case 'o':
          outfile = strdup (optarg);
          if (!fmt && outfile != NULL)
            {
#ifndef DISABLE_DXF
              if (strstr (outfile, ".dxf") || strstr (outfile, ".DXF"))
                fmt = strdup ("dxf");
              else if (strstr (outfile, ".dxfb") || strstr (outfile, ".DXFB"))
                fmt = strdup ("dxfb");
              else
#endif
#ifndef DISABLE_JSON
                  if (strstr (outfile, ".json") || strstr (outfile, ".JSON"))
                fmt = strdup ("json");
              else if (strstr (outfile, ".geojson")
                       || strstr (outfile, ".GeoJSON"))
                fmt = strdup ("geojson");
              else
#endif
                fprintf (stderr, "Unknown output format for %s\n", outfile);
            }
          break;
        case 'v': // support -v3 and -v
          i = (optind > 0 && optind < argc) ? optind - 1 : 1;
          if (!memcmp (argv[i], "-v", 2))
            {
              opts = argv[i][2] ? argv[i][2] - '0' : 1;
            }
          if (opts < 0 || opts > 9)
            return usage ();
#if defined(USE_TRACING) && defined(HAVE_SETENV)
          {
            char v[2];
            *v = opts + '0';
            *(v + 1) = 0;
            setenv ("LIBREDWG_TRACE", v, 1);
          }
#endif
          has_v = 1;
          break;
        case 'h':
          return help ();
        case '?':
          fprintf (stderr, "%s: invalid option '-%c' ignored\n", argv[0],
                   optopt);
          break;
        default:
          return usage ();
        }
    }
  i = optind;

  memset (&dwg, 0, sizeof (Dwg_Data));
  if (has_v || !fmt)
    dwg.opts = opts;
#if defined(USE_TRACING) && defined(HAVE_SETENV)
  if (!has_v)
    setenv ("LIBREDWG_TRACE", "1", 0);
#endif

#ifndef DISABLE_DXF
  if (optind != argc)
    {
      if (opts > 1)
        fprintf (stderr, "Reading DXF file %s\n", argv[i]);
      error = dxf_read_file (argv[i], &dwg);
    }
  else
    {
      if (opts > 1)
        fprintf (stderr, "Reading DXF from stdin\n");
      error = dxf_read_file ("-", &dwg); // i.e. from stdin
    }
#endif

  if (error >= DWG_ERR_CRITICAL)
    goto done;

  if (fmt)
    {
      Bit_Chain dat = { 0 };
      if (outfile)
        dat.fh = fopen (outfile, "w");
      else
        dat.fh = stdout;
      fprintf (stderr, "\n");
      dat.version = dat.from_version = dwg.header.version;
      dat.codepage = dwg.header.codepage;
      // TODO --as-rNNNN version? for now not.
      // we want the native dump, converters are separate.
#ifndef DISABLE_DXF
#  ifndef DISABLE_JSON
      if (!strcasecmp (fmt, "json"))
        {
          if (opts > 1 && outfile)
            fprintf (stderr, "Writing JSON file %s\n", outfile);
          error = dwg_write_json (&dat, &dwg);
        }
      else
#  endif
          if (!strcasecmp (fmt, "dxfb"))
        {
          if (opts > 1 && outfile)
            fprintf (stderr, "Writing Binary DXF file %s\n", outfile);
          error = dwg_write_dxfb (&dat, &dwg);
        }
      else if (!strcasecmp (fmt, "dxf"))
        {
          if (opts > 1 && outfile)
            fprintf (stderr, "Writing Binary DXF file %s\n", outfile);
          error = dwg_write_dxf (&dat, &dwg);
        }
#  ifndef DISABLE_JSON
      else if (!strcasecmp (fmt, "geojson"))
        {
          if (opts > 1 && outfile)
            fprintf (stderr, "Writing GeoJSON file %s\n", outfile);
          error = dwg_write_geojson (&dat, &dwg);
        }
      else
#  endif
#endif
        fprintf (stderr, "Invalid output format '%s'\n", fmt);

      if (outfile)
        fclose (dat.fh);
    }

done:
#if defined __SANITIZE_ADDRESS__ || __has_feature(address_sanitizer)
  {
    char *asanenv = getenv ("ASAN_OPTIONS");
    if (!asanenv)
      force_free = 1;
    // detect_leaks is enabled by default. see if it's turned off
    else if (strstr (asanenv, "detect_leaks=0") == NULL) /* not found */
      force_free = 1;
  }
#endif

  // forget about valgrind. really huge DWG's need endlessly here.
  if ((dwg.header.version && dwg.num_objects < 1000) || force_free
#ifdef HAVE_VALGRIND_VALGRIND_H
      || (RUNNING_ON_VALGRIND)
#endif
  )
    {
      if (fmt)
        free ((char *)fmt);
      if (outfile)
        free ((char *)outfile);
      dwg_free (&dwg);
    }

  if (error >= DWG_ERR_CRITICAL)
    {
      fprintf (stderr, "ERROR 0x%x\n", error);
      if (error && opts > 2)
        dwg_errstrings (error);
    }
  else
    {
      if (opts > 1)
        {
          fprintf (stderr, "SUCCESS 0x%x\n", error);
          if (error && opts > 2)
            dwg_errstrings (error);
        }
      else
        fprintf (stderr, "SUCCESS\n");
    }

  return error >= DWG_ERR_CRITICAL ? 1 : 0;
}
